File: Samples/ObjCSample/ObjCSample/NativeAdInTableViewController.m
//
//  NativeAdInTableTableViewController.m
//  ObjCSample
//
//  Created by Julien Gomez on 01/09/2015.
//  Copyright (c) 2015 Smart AdServer. All rights reserved.
//

#import "NativeAdInTableViewController.h"
#import "ListItem.h"
#import "RSSListItem.h"
#import "AdListItem.h"
#import "UIImageView+AFNetworking.h"
#import "DetailViewController.h"
#import "SASNativeAdManager.h"
#import "NewsItemCell.h"
#import "NativeAdCell.h"
#import "RatingView.h"

#define kIntervalBetweenNativeAds 4


@interface NativeAdInTableViewController () {
    NSMutableArray *_listItems;
    RSSFeedXMLParser *_XMLParser;
    BOOL _fullMode;
}
@end


@implementation NativeAdInTableViewController

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}


- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil fullMode:(BOOL)fullMode {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    
    if (self) {
        if (fullMode) {
            [self.tableView registerNib:[UINib nibWithNibName:@"NativeAdFullCell" bundle:nil] forCellReuseIdentifier:@"NativeAdCell"];
            [self.tableView registerNib:[UINib nibWithNibName:@"NewsItemFullCell" bundle:nil] forCellReuseIdentifier:@"NewsItemCell"];
        } else {
            [self.tableView registerNib:[UINib nibWithNibName:@"NativeAdCell" bundle:nil] forCellReuseIdentifier:@"NativeAdCell"];
            [self.tableView registerNib:[UINib nibWithNibName:@"NewsItemCell" bundle:nil] forCellReuseIdentifier:@"NewsItemCell"];
        }
        _fullMode = fullMode;
    }
    
    return self;
}


- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    return [self initWithNibName:nibNameOrNil bundle:nibBundleOrNil fullMode:NO];
}


- (void)viewDidLoad {
    [super viewDidLoad];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(updateTableView:) name:kNativeAdReadyNotification object:nil];
    
     self.title = @"Native Ad";
    
    // Initialize the refresh control.
    self.refreshControl = [[UIRefreshControl alloc] init];
    self.refreshControl.backgroundColor = [UIColor purpleColor];
    self.refreshControl.tintColor = [UIColor whiteColor];
    [self.refreshControl addTarget:self
                            action:@selector(refresh)
                  forControlEvents:UIControlEventValueChanged];
    
    _listItems = nil;
    [self refresh];
}


- (void)refresh {
    [self loadRSSFeed:[NSURL URLWithString:@"http://blog.smartadserver.com/feed/"]];
}

#pragma mark - RSS Request Manager

- (void)loadRSSFeed:(NSURL *)feedURL {
    if(_XMLParser == nil) {
        _XMLParser = [[RSSFeedXMLParser alloc] init];
        _XMLParser.delegate = self;
    }
    [_XMLParser startParsingWithContentsOfURL:feedURL];
}


- (void) xmlParser:(RSSFeedXMLParser *)parser didDownloadRSSItems:(NSArray *)rssItems withFeedURL:(NSURL *)feedURL {
   
    _listItems = [NSMutableArray array];
    
    for (int i = 0; i < [rssItems count]; i++) {
        
        //Every kIntervalBetweenNativeAds items, insert a native ad in the list
        if (i % kIntervalBetweenNativeAds == 0) {
            
            // The config object is used to represent the placement of the banner.
            // The information provided are almost the same than the information you need to provide to a classic display banner,
            // except for the 'master' parameter that does not exist for native ads.
            SASNativeAdPlacement *adPlacement = [SASNativeAdPlacement nativeAdPlacementWithBaseURL:[NSURL URLWithString:@"http://mobile.smartadserver.com"] siteID: 28298 pageID:@"586114" formatID:15140 target:@""];

            AdListItem *adItem = [[AdListItem alloc] initWithAdPlacement:adPlacement];
            [_listItems addObject:adItem];
        }
        
        RSSListItem *rssListItem = rssItems[i];
        [_listItems addObject:rssListItem];
    }
    
    [self.tableView reloadData];
    [self.refreshControl endRefreshing];

}


#pragma mark - Table View

- (void)updateTableView:(NSNotification *)notification {
    [self.tableView beginUpdates];
    
    //Reload all cells containing ads (one cell every kIntervalBetweenNativeAds cells)
    NSMutableArray *indexPaths = [NSMutableArray array];
    for (int i = 0; i < [_listItems count]; i = i + 1 + kIntervalBetweenNativeAds) {
        [indexPaths addObject:[NSIndexPath indexPathForRow:i inSection:0]];
    }
    [self.tableView reloadRowsAtIndexPaths:indexPaths withRowAnimation:UITableViewRowAnimationAutomatic];
    
    [self.tableView endUpdates];
}


- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}


- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat height = 70;
    if (_fullMode) {
        height = 220;
    }
    
    if ([_listItems[indexPath.row] isKindOfClass:[AdListItem class]]) {
        AdListItem *item = _listItems[indexPath.row];
        if ([item isReady]) {
            return height;
        } else {
            return 0;
        }
    } else {
        return height;
    }
}


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _listItems.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *NewsItemCellIdentifier = @"NewsItemCell";
    static NSString *AdCellIdentifier = @"NativeAdCell";
    
    UITableViewCell *cell;
    
    if ([_listItems[indexPath.row] isKindOfClass:[AdListItem class]]) {
        AdListItem *item = _listItems[indexPath.row];
        NativeAdCell *adCell = (NativeAdCell *)[tableView dequeueReusableCellWithIdentifier:AdCellIdentifier];
        if ([item isReady]) {
            if (adCell.adItem) {
                // Unregister previous ad if needed
                [adCell.adItem unregisterViews];
            }
            adCell.adItem = item;
        
            adCell.adTitleLabel.text = item.title;
            adCell.adDescriptionLabel.text = item.subtitle;
   
            [adCell.adCallToActionButton setTitle:item.callToAction forState:UIControlStateNormal];
            [adCell.adIconImageView setImageWithURL:item.imageURL placeholderImage:[UIImage imageNamed:@"NewsIcon.png"]];
            [adCell.adCoverImageView setImageWithURL:item.coverImageURL placeholderImage:nil];
            
            UIView *view = (UIView *)[adCell viewWithTag:100];
            UIView *previousRatingView = (UIView *)[adCell viewWithTag:101];
            if (previousRatingView) {
                [previousRatingView removeFromSuperview];
            }
            
            if (item.rating != SASRatingUndefined) {
                RatingView *ratingView = [[RatingView alloc] initWithFrame:CGRectMake(0, 0, 55, 10) andRating:item.rating];
                ratingView.tag = 101;
                
                [view addSubview:ratingView];
            }
            
            [item registerView:adCell modalParentViewController:self];
        
            [cell setHidden:NO];
        } else {
            [cell setHidden:YES];
        }
        
        cell = (UITableViewCell *)adCell;
    } else {
        RSSListItem *item = _listItems[indexPath.row];
        NewsItemCell *newsItemCell = [tableView dequeueReusableCellWithIdentifier:NewsItemCellIdentifier];
        newsItemCell.newsItemTitleLabel.text = item.title;
        newsItemCell.newsItemDescriptionLabel.text = item.subtitle;
        [newsItemCell.newsItemIconImageView setImageWithURL:item.imageURL placeholderImage:[UIImage imageNamed:@"NewsIcon.png"]];
        newsItemCell.publishedDateLabel.text = item.publishedDate;
       
        cell = (UITableViewCell *)newsItemCell;
    }
    
    
    return cell;
}


- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    if ([_listItems[indexPath.row] isKindOfClass:[RSSListItem class]]) {
        DetailViewController *controller = [[DetailViewController alloc] init];
        [controller setListItem:_listItems[indexPath.row]];
        [self.navigationController pushViewController:controller animated:YES];
    }
}


@end

File: Samples/SwiftSample/SwiftSample/NativeAdInTableViewController.swift
//
//  NativeAdInTableViewController.swift
//  SwiftSample
//
//  Created by Julien Gomez on 03/09/2015.
//  Copyright © 2015 Smart AdServer. All rights reserved.
//

import UIKit


// Example of a native ad integration in an existing table view.
//
// For this integration, we choose to insert a native ad every X items (this number can be set through a variable),
// but it can work the same way with fixed positioned ad banners.
//
// This sample relies on a list of ListItem that can be from to types:
//   - RSSListItem: business object representing an actual content item from your app
//   - AdListItem: an ad object that will load and represents a native ad.
//
// The controller will first load the business items from a RSS feed in this example, then insert AdListItem objects
// in the item list.
// Each AdListItem will instantiate a native ad manager and load its own ad, then warn the controller using the notification.
class NativeAdInTableViewController: UITableViewController, XMLParserDelegate {
    
    // This notification is used to refresh the tableview cells containing ads when they are updated by their respective ad manager
    static let NATIVE_AD_READY_NOTIFICATION = "NativeAdReadyNotification"
    
    // The interval between native ads in the table view
    private let INTERVAL_BETWEEN_NATIVE_ADS = 4

    // Other objects to handle the RSS parsing and the 'article' display
    private let xmlParser = XMLParser()
    private var items = [ListItem]()
    let postClickViewController = NativeAdPostClickViewController(nibName: "NativeAdPostClickViewController", bundle: nil)
    
    // MARK: - View controller lifecycle
    
    deinit {
        // Do not forget to remove your controller from the notification center when it is deallocated
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        retrieveRSSFeed()
        createReloadButton()
        
        // Add a notification observer to know when an ad cell needs to be refreshed
        NSNotificationCenter.defaultCenter().addObserver(self, selector: Selector("updateTableView:"), name: NativeAdInTableViewController.NATIVE_AD_READY_NOTIFICATION, object: nil)
    }
    
    private func retrieveRSSFeed() {
        let url = NSURL(string: "http://blog.smartadserver.com/feed/")
        xmlParser.delegate = self
        xmlParser.startParsingWithContentsOfURL(url)
    }
    
    func createReloadButton() {
        let loadButton = UIBarButtonItem(title: "Reload", style: .Bordered, target: self, action: "reload")
        navigationItem.rightBarButtonItem = loadButton
    }
    
    func reload() {
        items.removeAll()
        self.tableView.reloadData()
        
        retrieveRSSFeed()
    }

    // MARK: - XMLParserDelegate method implementation
    
    func parser(parser: XMLParser, didFinishParsing results: [Dictionary<String, String>]) {
        items.removeAll()
        
        // Each business item retrieved through the RSS feed is converted in a RSSListItem and inserted in the items list
        var i = 0
        for result in results {
            
            // Every X items, an AdListItem is inserted
            if (i % INTERVAL_BETWEEN_NATIVE_ADS == 0) {
                
                // The config object is used to represent the placement of the banner.
                // The information provided are almost the same than the information you need to provide to a classic display banner,
                // except for the 'master' parameter that does not exist for native ads.
                let adPlacement = SASNativeAdPlacement(baseURL: NSURL(string: Constants.baseURL())!, siteID: Constants.siteID(), pageID: Constants.nativeAdInTableViewPageID(), formatID: Constants.nativeAdFormatID(), target: "")
                
                // The ad item is created with the configuration and added to the items list
                let adItem = AdListItem(adPlacement: adPlacement)
                items.append(adItem)
            }
            
            if let title = result[XMLParser.TITLE_KEY], let subtitle = result[XMLParser.DESCRIPTION_KEY], let image = result[XMLParser.URL_KEY], let link = result[XMLParser.LINK_KEY] {
                let RSSItem = RSSListItem(title: title, subtitle: subtitle, image: image, link: link)
                items.append(RSSItem)
            }
            
            ++i
        }
        
        self.tableView.reloadData()
    }
    
    // MARK: - Table view data source
    
    func updateTableView(notification: NSNotification) {
        tableView.beginUpdates()
        
        // When an ad notification is received, ad cells needs to be reloaded.
        var indexPaths = [NSIndexPath]()
        for var i = 0; i < items.count; i = i + 1 + INTERVAL_BETWEEN_NATIVE_ADS {
            indexPaths.append(NSIndexPath(forRow: i, inSection: 0))
        }
        tableView.reloadRowsAtIndexPaths(indexPaths, withRowAnimation: UITableViewRowAnimation.Automatic)
        
        tableView.endUpdates()
    }

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return items.count
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cellId = items[indexPath.row] is AdListItem ? "NativeAdCell" : "NewsItemCell"
        
        let cell = tableView.dequeueReusableCellWithIdentifier(cellId, forIndexPath: indexPath) as! ListItemCell
        
        if let title = cell.viewWithTag(1) as? UILabel, let subtitle = cell.viewWithTag(2) as? UILabel, let image = cell.viewWithTag(3) as? UIImageView {
            
            // Since an ad and a RSS item are representing using the same kind of class, it is easy to fill a cell for a given row.
            let item = items[indexPath.row]
            title.text = item.title()
            subtitle.text = item.subtitle()
            image.setImageWithURL(NSURL(string: item.image()), placeholderImage: UIImage(named: "NewsIcon.png"))
            image.contentMode = UIViewContentMode.ScaleAspectFit
            
            // Unregister previous ad if needed
            if let previousItem = cell.adItem {
                previousItem.unregisterViews()
            }
            
            // IMPORTANT: you have to indicate clearly that the item is an ad to comply with the law in most countries.
            //            Please check your local legislation for more details.
            if let item = (item as? AdListItem) where item.isReady(), let button = cell.viewWithTag(4) as? UIButton {
                button .setTitle(item.callToAction(), forState: UIControlState.Normal)
                cell.adItem = item
                
                // The views used by the native ads must be registered on the native ad as soon as they are ready and displayed.
                // This will ensure that the impression is counted and will handle click automatically.
                // Registered views must be unregistered using the unregisterViews method before being released or reuse for another
                // ad (be careful with table view cells).
                item.registerView(cell, modalParentViewController: self)
            } else {
                cell.adItem = nil
            }
        }
        
        return cell
    }
    
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        let item = items[indexPath.row]
        if item is AdListItem && !(item as! AdListItem).isReady() {
            // Hide the ad cell if it's not ready
            return 0
        } else {
            return 70
        }
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        
        let item = items[indexPath.row]
        if item is RSSListItem {
            // Since the click is handled automatically for native ads, you need to check that the cell clicked is
            // a RSS item and not an ad before displaying the article.
            postClickViewController.item = item as? RSSListItem
            navigationController?.pushViewController(postClickViewController, animated: true)
        }
    }

}
